Add initial spec for OONI run (#249)
ooni/spec: backends/bk-005-ooni-run.md

# OONI Run specification

* author: Norbel Ambanumben
* version: 0.0.1
* date: 2022-06-15
* status: _draft_

> This document aims at providing a functional specification for OONI Run.

Reviewer: ```suggestion
This document provides a functional specification for OONI Run.
```

# 1.0 System overview

> The api exposes an HTTP API allowing OONI Run web app to submit links to the

Reviewer: ```suggestion
The ooni/api exposes an HTTP API allowing OONI Run web app to submit links to the
```

backend and also allow the OONI Probe to query OONI Run link data.
The client and the server MUST NOT assume a keep
alive semantics for the HTTP connections.

> 

Reviewer 2: I think we should add a list of definitions for what we mean when we talk about "OONI Run link" and "OONI Run link data".

I suggest defining the following terms:
- OONI Run link - This is the mobile deep link that when a user taps on it having the OONI Probe app installed, will trigger opening the app and be handled by it. If it's accessed from a device that doesn't have the app installed (or the handling is broken) it will display a web page prompting the user to install the app or allowing them to force opening the app with the custom scheme `ooni://`
- OONI Run descriptor - This is the metadata for an experiment run
- OONI Run descriptor URL - This is the URL used to retrieve the descriptor from a backend

New server implementations MUST support `Content-Encoding: gzip` to
compress request bodies. OONI production clients MUST NOT use this
feature until all OONI's production servers server are known to have
implemented it. Third party clients should either coordinate with

> us or be prepared to retry without compression.

Reviewer 2: This is not relevant to this OONI Run.

New implementations MUST properly set `Content-Type`. Server side
implementatons MUST be able to deal with legacy clients that possibly

> do not correctly set the `Content-Type`.

Reviewer 2: This is not relevant to OONI Run. There is no notion of "New" implementations since we never had a backend for OONI Run.

> Tests submitted to the OONI Run API will be stored and processed by OONI backend
> and OONNI Probes the URL is shared with.
> How that will happen is out of the scope of this document. You may also want to read
> [about our data policy](https://ooni.io/about/data-policy/).

Reviewer: ```suggestion
Tests submitted to the OONI Run API will be stored and processed by OONI backend
and OONI Probes the URL is shared with.
How that will happen is out of the scope of this document. You may also want to read
[about our data policy](https://ooni.io/about/data-policy/).
```

That said, I think this paragraph is not very relevant for this specification.

The OONI Run MUST be exposed as an HTTPS service. It MUST also be exposed as

> a Tor onion service as long as legacy OONI probe clients use it. The need

Reviewer: There are no legacy OONI Probe clients using OONI Run. So, this paragraph should probably be simplified.

to expose an onion service will be rediscussed when legacy OONI probe
clients will no longer be relevant. A [legacy document](
https://ooni.torproject.org/docs/architecture.html)
explains why the OONI project originally chose to allow for both HTTPS
and Tor onion service services (henceforth, Onion).

It is also outside of the scope of this section to define the way in which
a OONI probe discovers the collector API endpoint, as well as how, given
several endpoints and/or endpoint types (HTTPS, Onion), it chooses a specific
endpoint.

> 

Reviewer 2: I think more details need to be provided on the high level architecture of how this specific system is going to work. Specifically we need to document how we go from handling an OONI Run link to retrieving the OONI Run descriptor via the OONI Run descriptor URL

> endpoint.

Reviewer: Please remove this paragraph. It's about the collector.

# 2.0 Threat model

The OONI RUN Service transport MUST guarantee some reasonable level of encryption
and authentication between the OONI probe and itself. Therefore, a malicious
lazy actor won't be able to easily modify test results that they have not
created while they are being submitted to the OONI Run Service. Also, they will not

> be able to easily log the messages exchanged with the OONI Run Service.

Reviewer: Please change the wording. This wording clearly still refers to the collector.

It is outside of the scope of the OONI Run Service to provide blocking resistance or
to conceal to a passive network observer the fact that they are communicating to
a OONI Run Service. Such properties are to be provided by other software, e.g. Tor.

Therefore a client implementation of the OONI Run Service protocol SHOULD allow one
to specify a [SOCKS5](https://tools.ietf.org/html/rfc1928) proxy where the
name resolution is performed by the circumvention tool (`socks5h`).

# 3.0 API

> The same API is available via HTTP and Onion.

Reviewer 2: I don't think we should be exposing this via onion service.

The standard flow for OONI RUN Link is the following:

> - you create an OONI RUN Link (see 3.1);

Reviewer 2: I think it's important here to clarify the various components you are interacting with when creating an OONI Run link and what are the bits of information that you get back when creating the link.

For example at some point you are interacting with a web service that speak to an API backend and the thing you get back is an OONI Run link that when opening with the mobile app speaks to the API backend to retrieve the OONI Run descriptor.

- you update it by adding more tests or metadata (see 3.2);

- you access it from probes or frontend service (see 3.3);

- then, you can disable/delete it (see 3.4).

## 3.1 Create a new OONI Run link

### Request

> When a you *create* a new OONI RUN link, the frontend sends a `POST`

Reviewer: ```suggestion
When you *create* a new OONI RUN link, the frontend sends a `POST`
```

request conforming to the following, informal specification:

>     POST /link

Reviewer: ```suggestion
    POST /api/v1/oonirun/link
```

>     POST /api/v2/oonirun/link

Reviewer 2: I think this should still be nested under `/api/v1/` since the versioning is of the OONI API not the OONI Run link itself. As we don't have a `v1` OONI Run link speaking to the API we can use this without any issues.

I suggest we use as an endpoint:
```
/api/v2/oonirun/link
```

    {
     "name":
        (required) `string` it is required property used in identifying the link in the OONI probe and other parts of the system.

     "description":
        (optional) `string` describing the tests included as part of the OONI Run link.
      
      "tests": `array` provides a JSON array of tests to be run.
         [
            {
                  "ta": {

>                      "urls": []

Reviewer: ```suggestion
                     "inputs": []
```

We're already cursed enough by the fact that we think the input the experiments are URLs. Let's be more generic and allow more flexibility. It will pay off in the future.

(I know the v1 format says `urls` but this is v2 and we should make it better, so...)

                  },
                  "tn": "web_connectivity"
            },
            {
                  "tn": "telegram"
            },
            {
                  "tn": "signal"
            }
         ]
    }

Where the above is intended as an informal specification for generating a
compliant, serialized JSON object.

### Response status code

Upon receiving a request to create a report, the OONI Run Service:

1. SHOULD fail with `4xx` if the request body does not parse, it is not a JSON object,
   any required field is missing and/or if any present field has an invalid value.

> 2. MUST fail with `4xx` if the request is not compliant with its policies.

Reviewer: Which policies?

3. MUST fail with `5xx` if it cannot generate the OONI Run ID (see below) or
   in case of other failures opening the OONI Run.

4. if everything is okay, MUST return a `200` response.

### Response body

In case of failure, the OONI Run Service MUST return a JSON object, whose content
is implementation dependent and MAY be empty. Some existing implementations
return a `{"error": "string"}` object to make debugging easier.

In case of success (i.e. `200` response), the OONI Run Service MUST return a JSON body
generated in compliance with the following, informal specification:

    {
      "backend_version":
        `string` containing the version of the backend. This version MUST

>         match the regex provided above for 'software_version'.

Reviewer: I am not a fan of returning this information. We used to do that in the collect for legacy reasons. I think we may also decide to avoid returning this information, actually.

      "ooni_run_id":
        `string` OONI Run identifier. The format of this field is not
        specified, except that is MUST be a valid UTF-8 string, of
        course. The client MUST NOT make any assumption with respect
        to the structure of this field.
    }

### Response processing requirements

Upon receiving a response, new written clients MUST check that the
status is `200` before continuing, MUST ensure that the server supports

> their preferred data submission format, and MUST save the OONI Run ID.

Reviewer: Please edit to remove unnecessary wording borrowed from the collector spec

## 3.2 Update an OONI Run Link

> Updating an OONI Run Link means appending a test to the OONI Run Link.

Reviewer: What about adding new URLs?

### Request

To update an OONI Run Link, the client issues a request compliant with:

>     POST /report/${ooni_run_id}

Reviewer: ```suggestion
    POST /api/v1/oonirun/link/${ooni_run_id}
```

    {
     "name":
        (required) `string` it is required property used in identifying the link in the OONI probe and other parts of the system.

     "description":
        (optional) `string` describing the tests included as part of the OONI Run link.
      
      "tests": `array` provides a JSON array of tests to be run.
         [
            {
                  "ta": {

>                      "urls": []

Reviewer: ```suggestion
                     "inputs": []
```

Same rationale as above

                  },
                  "tn": "web_connectivity"
            },
            {
                  "tn": "telegram"
            },
            {
                  "tn": "signal"
            }
         ]
    }

### Response status code

> Upon receiving this request, the collector:

Reviewer: ```suggestion
Upon receiving this request, the OONI Run backend:
```

1. SHOULD check whether `${ooni_run_id}` is a valid, OONI Run ID and reject
   the request with a `4xx` status otherwise.

2. SHOULD reject the request with a `4xx` if the JSON does not
   parse or the parsed value is not a JSON object.

3. MUST "commit" the OONI Run data to persistent storage or to some
   database before returning `200` to the client making
   sure that it successfully saved the OONI Run Data (e.g. by
   checking the return value of `fclose`). There SHOULD be
   integration tests to check whether the JSON serialized by
   a specific implementation is compliant with the format
   expected by the pipeline (e.g. in `go` with `omitempty`
   `null` values are removed by a marshal-serialize cycle

>    while the Python parser preserves them).

Reviewer: I don't think we need to be so strict. This wording was very careful because we were describing measurement submission but I don't think there are the same requirements here.

4. is allowed to perform additional quick operations that may have
   an impact on the status code, and SHOULD defer other operations
   that do not have an impact on the status code to after the status
   code has been sent to the client (the goal being to keep the
   connection open for as little as possible to avoid the risk that
   a middlebox in a constrained network flags the connection as

>    stale and closse it, thus preventing the server from telling the client

Reviewer: ```suggestion
   stale and close it, thus preventing the server from telling the client
```

>    that the measurement has been successfully submitted).

Reviewer: Not sure if this is needed here

5. if everything is okay, returns `200` to the client (see below).

### Response body

In case of failure, the OONI Run Service MUST return a JSON object, whose content
is implementation dependent and MAY be empty. Some existing implementations
return a `{"error": "string"}` object to make debugging easier.

In case of `200` responses, new OONI Run Service implementations MAY
include a unique identifier for the OONI Run Link that the
client may later use to reference said OONI Run Link. For example:

```JSON
{"ooni_run_id":"e00c584e6e9e5326"}

> ```

Reviewer: I'm not sure about returning the ID here. We already know it because we just modified the OONI Run link.

As far as the client is concerned, this `ooni_run_id` is an
opaque UTF-8 string that has some meaning to the server.

## 3.3 Deleting an OONI Run Link

To delete a report, a client should submit a request like:

>     DELETE /link/${ooni_run_id}

Reviewer: ```suggestion
    DELETE /api/v1/oonirun/link/${ooni_run_id}
```

If the report exists, `200` is returned
and the response MUST include this body for backwards
compatibility with existing implementations:

```JSON
{"status": "success"}
```

Otherwise, in case of failure, the OONI Run Service MUST return a JSON object, whose

> content is implementation dependent and MAY be empty. Some existing implementations

Reviewer: Please mention that it should return `4xx` on failure

return a `{"error": "string"}` object to make debugging easier.

# 4.0 Implementation considerations

A client side implementation of the OONI Run Service MUST make sure

> that it is emitting timestamps using UTC rather than local time.

Reviewer: Not sure this is relevant

A client side implementation MAY retry any failing OONI Run Service operation
immediately for three times in case there is a DNS or TCP error. This
is to ensure that transient errors do not prevent us from submitting the
data immediately. If all these immediate retries fail, then the client
SHOULD arrange for resubmitting the OONI Run Link at a later time, either
requiring user input or automatically. In the latter case, the delay
after which the client will attempt to resubmit SHOULD be exponentially

> distributed and SHOULD NOT be smaller than 15 minutes.

Reviewer: Not sure this is relevant

A server implementation MAY publish metrics allowing OONI to gradually
enforce more JSON schema correctness in the OONI Run Service, both for the
open-report request and for submitted data. A possible strategy to make
this happen consists of parsing without rejecting, counting the number
of failures, and then iterating until the number of failures starts
converging to zero. At that point, it is possible to make the checks

> mandatory and reject invalid input.

Reviewer: Not sure this is relevant

# OONI Run specification

* author: Norbel Ambanumben, Arturo Filastò

> * version: 2023.02.09

Reviewer: I removed version. We already have date and we can format dates to be version numbers, like I did. It is already burdensome to maintain one identifier, so we should not maintain two. We should migrate all specs to have their specification version number being the date of the last modification because there is no point in saying "I am implementing version 0.0.1 of oonirun v2 specification. We always implement the last version and we already factor in breaking changes by changing the name of the spec. For example, `ooni-run` and `ooni-run-v2`. In this vein, I am going to clearly name the file such that it's obvious we're referring to `ooni-run-v2`.

      },
      {
         "test_name": "dnscheck"
      }
   ]
}
```

# 5.0 Implementation considerations

Special attention should be placed in ensuring the OONI Run links (which are
mobile deep links) are sharable though various apps.
In particular the format of the OONI Run link should be such that if the URL is
being truncated, it should be visible to the end, as such it's recommend that we
restrict the character set of the `ooni_run_link_id` to just numbers. Since we
might not end up having that many OONI Run link, this also lends itself well to
allowing users to manually type OONI Run links directly into the app.

Mobile deep links can be registered using two different methods, one is a custom
prefix (ex. `ooni://`), the other is a custom URL prefix (ex.
`https://run.ooni.io/1234`). In our testing we have seen that the custom prefix
is more reliable, yet it has the tradeoff of not allowing us to display a web
page when the user does not have the app installed. As such the recommended
strategy is to encourage users to share the custom URL prefix OONI Run link, but
on the web page itself, in the event that the app did not handle the deep link,

> have a link to the custom prefix approach to "force" the opening of the app.

Reviewer: I think we should clarify the last sentence. Does it mean that the user will have a webpage with a link and the _option_ of clicking on the link to trigger the app?

Reviewer 2: Yes that's correct. Do you have a suggestion on how to improve the text? I added a sentence at the end saying this is how OONI Run links currently work.

As such we recommend using the following addresses for OONI Run link and OONI Run descriptor URLs:

* `https://run.ooni.io/{ooni_run_link_id}`, where `{ooni_run_link_id}` is a number

* `ooni://runv2/{ooni_run_link_id}`

* `https://api.ooni.io/api/v1/ooni_run/{ooni_run_link_id}`

Moreover, we could at some point host these links on s3 or github and have them

> be accessible via URLs in the form:

Reviewer: I think it is unclear how authentication and installation would work in this case. I suggest we move this last portion of the document into its own section, named "Future work".

Reviewer 2: Moved
